plutus-ledger-api/src/PlutusLedgerApi/Common/Versions.hs

```diff
@@ -153,44 +104,173 @@ ledgerLanguageIntroducedIn = \case
+'MajorProtocolVersion'?  See Note [New builtins/language versions and protocol
+versions].  This function (and others in this module) assumes that once a LL is
+available it remains available in all later PVs and that if m <= n, PlutusVm is
+introduced no later than PlutusVn.
 -}
 ledgerLanguagesAvailableIn :: MajorProtocolVersion -> Set.Set PlutusLedgerLanguage
 ledgerLanguagesAvailableIn searchPv =
-    foldMap ledgerVersionToSet enumerate
-  where
-    -- OPTIMIZE: could be done faster using takeWhile
-    ledgerVersionToSet :: PlutusLedgerLanguage -> Set.Set PlutusLedgerLanguage
-    ledgerVersionToSet ll
-        | ledgerLanguageIntroducedIn ll <= searchPv = Set.singleton ll
-        | otherwise = mempty
+  Set.fromList $ takeWhile (\ll -> ledgerLanguageIntroducedIn ll <= searchPv) enumerate
 
-{-| Which Plutus Core language versions are available in the given 'PlutusLedgerLanguage'
-and 'MajorProtocolVersion'?
+-- | Given a map from PVs to a type `a`, return a `Set a` containing all of the
+-- entries with PV <= thisPv
+collectUpTo
+  :: Ord a
+  => Map.Map MajorProtocolVersion (Set.Set a)
+  -> MajorProtocolVersion
+  -> Set.Set a
+collectUpTo m thisPv =
+  fold $  -- ie, iterated `union`
+  Map.elems $ Map.takeWhileAntitone (<= thisPv) m
 
-See Note [New builtins/language versions and protocol versions]
+{- Batches of builtins which were introduced in the same hard fork (but perhaps
+   not for all LLs): see the Plutus Core specification and
+   `builtinsIntroducedIn` below.
 -}
-plcVersionsAvailableIn :: PlutusLedgerLanguage -> MajorProtocolVersion -> Set.Set Version
-plcVersionsAvailableIn thisLv thisPv = fold $ Map.elems $
-    Map.takeWhileAntitone plcVersionAvailableIn plcVersionsIntroducedIn
-    where
-      plcVersionAvailableIn :: (PlutusLedgerLanguage, MajorProtocolVersion) -> Bool
-      plcVersionAvailableIn (introducedInLv,introducedInPv) =
-          -- both should be satisfied
-          introducedInLv <= thisLv && introducedInPv <= thisPv
 
-{-| Which builtin functions are available in the given given 'PlutusLedgerLanguage'
-and 'MajorProtocolVersion'?
+{- If any new builtins are introduced after a batch has been deployed on the chain
+  then a new `batch` object MUST be added to contain them and the
+  `builtinsIntroducedIn` function must be updated.  Also, remember to UPDATE THE
+  TESTS in `Spec.Versions` and `Spec.Data.Versions` when a new batch is added.
+-}
 
-See Note [New builtins/language versions and protocol versions]
+{- It's tempting to try something like `fmap toEnum [0..50]` here, but that's
+   dangerous because the order of the constructors in DefaultFun doesn't
+   precisely match the order that the builtins were introduced in.  A safer
+   alternative would be to use the flat tags, but they're not directly
+   accessible at the moment.
+-}
+batch1 :: [DefaultFun]
+batch1 =
+  [ AddInteger, SubtractInteger, MultiplyInteger, DivideInteger, QuotientInteger
```

Comment:
It's tedious to check the correctness of these batches manually. How about keeping the old version of `builtinsIntroducedIn`, and adding some property tests to verify that the old and new versions always agree?

Reply:
I'm not sure what you mean.  Do you want to keep both the old and the new versions?  The batches were implcitly [present](https://github.com/IntersectMBO/plutus/blob/2a70305402e563493d4cc6c195a573f6c5f9c529/plutus-ledger-api/src/PlutusLedgerApi/Common/Versions.hs#L90) in the previous version and here I've pulled them out to the top level to make things a bit more explicit and easier to relate to the specification; also, they're used in the tests so it helps to be able to export them.  Of course, you should only ever add anything to the most recent batch (or add a new batch); maybe I should add a prominent comment to point that out.

Reply:
> So we'd have to keep the old builtinsAvailableIn function around specifically for the tests, even though it's not used?

Yeah, why not, we can move it to a test package. But if there's already a test that verifies that the new `builtinsAvailableIn` gives correct results for all ll/pv (especially PV10 and PV11), then we are good - that coverage is all we need.

Reply:
OK.  The old version works correctly up to PV10, so I can move it into the test code somewhere and add a test that the new implementation matches up to PV10 (but not PV11, where it breaks).